sql/sql_statistics.cc

```diff
@@ -1580,6 +1594,140 @@ class Histogram_builder
   }
 };
 
+class Histogram_builder_json : public Histogram_builder
+{
+std::vector<std::string> bucket_bounds = {};
+
+public:
+  Histogram_builder_json(Field *col, uint col_len, ha_rows rows)
+  : Histogram_builder(col, col_len, rows)
+  {
+    Column_statistics *col_stats= col->collected_stats;
+    min_value= col_stats->min_value;
+    max_value= col_stats->max_value;
+    histogram= &col_stats->histogram;
+    hist_width= histogram->get_width();
+    bucket_capacity= (double) records / (hist_width + 1);
+    curr_bucket= 0;
+    count= 0;
+    count_distinct= 0;
+    count_distinct_single_occurence= 0;
+  }
+
+  ~Histogram_builder_json() override = default;
+
+  int next(void *elem, element_count elem_cnt) override
+  {
+    count_distinct++;
+    if (elem_cnt == 1)
+      count_distinct_single_occurence++;
+    count+= elem_cnt;
+    if (curr_bucket == hist_width)
+      return 0;
+    if (count > bucket_capacity * (curr_bucket + 1))
+    {
+      column->store_field_value((uchar *) elem, col_length);
+      StringBuffer<MAX_FIELD_WIDTH> val;
+      column->val_str(&val);
+      bucket_bounds.emplace_back(val.c_ptr());
+      curr_bucket++;
+    }
+    return 0;
+  }
+
+  void build_json_from_histogram() {
+    Json_writer *writer = new Json_writer();
+    writer->start_array();
+    for(auto& value: bucket_bounds) {
+      writer->add_str(value.c_str());
+    }
+    writer->end_array();
+    histogram->set_size(bucket_bounds.size());
+    Binary_string *json_string = (Binary_string *) writer->output.get_string();
+    histogram->set_values((uchar *) json_string->c_ptr());
+
+    std::vector<std::string> buckets = parse_histogram_from_json(json_string->c_ptr());
+    printf("%zu", buckets.size());
+
+    test_parse_histogram_from_json();
+  }
+
+  static std::vector<std::string> parse_histogram_from_json(const char *json)
+  {
+    std::vector<std::string> hist_buckets= {};
+    enum json_types vt = json_get_array_items(json, json + strlen(json), hist_buckets);
+    printf("%d", vt);
+    printf("%zu", hist_buckets.size());
+
```

Comment:
I've asked to print the diagnostics to stderr, where one can see it!  
So ,I have modified this patch to print to stderr, and also to print more:
https://gist.github.com/spetrunia/16f4a410b4702453f7ea7b2bb43abb7e

Then, I run this test:
```
create table ten(a int primary key);
insert into ten values (0),(1),(2),(3),(4),(5),(6),(7),(8),(9);
set histogram_type=json;
analyze table ten persistent for all;
```

and I see this output:
```
  ==============
  histogram: [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9"
  ]
  json_get_array_items() returned -1
   JSV_BAD_JSON=-1, JSON_VALUE_ARRAY=2
  hist_buckets.size()=0
```

The histogram looks ok but the parsing didn't succeed?  Please fix this.

Reply:
Thanks for the patch! I tried running with `./mysql-test/mtr ...` but couldn't get the output (from `fprintf`) to show up. Do you mind sharing how you run it from your end?

Reply:
mtr saves the server error log in the .err file.
```
cd mysql-test
./mtr statistics_json
vim var/log/mysqld.1.err
```